streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java

```diff
@@ -31,60 +31,71 @@
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.regex.Pattern;
 
+import static org.apache.kafka.streams.kstream.internals.StreamsTopologyGraphImpl.TOPOLOGY_ROOT;
+import static org.apache.kafka.streams.kstream.internals.TopologyNodeType.GLOBAL_KTABLE;
+import static org.apache.kafka.streams.kstream.internals.TopologyNodeType.KTABLE;
+import static org.apache.kafka.streams.kstream.internals.TopologyNodeType.SOURCE;
+
+@SuppressWarnings("unchecked")
 public class InternalStreamsBuilder implements InternalNameProvider {
 
     final InternalTopologyBuilder internalTopologyBuilder;
```

Comment:
If we alter the `add*StateStore()` methods to put the state stores in the `StreamsTopologyGraph`, this class could actually avoid holding a reference to `InternalTopologyBuilder` at all. It would then become the sole responsibility of the `TopologyOptimizer` to use the `InternalTopologyBuilder`.

In fact, that would also let you drop `buildAndOptimize` and `topologyNeedsBuilding`. `org.apache.kafka.streams.StreamsBuilder#build` would instead just `getTopologyGraph` and pass it to the `TopologyOptimizer`.

Now that I'm looking at this after our chat this morning, It was this reference that made me think these classes were more coupled than they are.

Reply:
Ack